system/jlib/jsocket.cpp

```diff
@@ -5183,24 +5204,37 @@ class CSocketEpollHandler: implements ISocketSelectHandler, public CInterface
     void remove(ISocket *sock)
     {
         CriticalBlock block(sect);
-        unsigned nt = threads.ordinality();
-        ForEachItemIn(i,threads)
+        unsigned nt = threads.size();
+        bool removed = false;
+        std::list<Owned<CSocketEpollThread>>::iterator it = threads.begin();
+        while (it != threads.end())
         {
-            if (threads.item(i).remove(sock, nt))
-                break;
+            if (!removed)
+            {
+                if ((*it)->remove(sock, nt) && sock)
+                    removed = true;
+            }
+            if ( ((*it)->finished) && ((*it)->join(0)) )
```

Comment:
The join and delete should really be outside the critical section, just in case they take a while.

Reply:
I thought should maintain serialization for reading/writing to the threads list.
join(0) has arg of zero and is only called if finished so that it did not block.